scripts/cluster-install/install.py

```diff
@@ -0,0 +1,392 @@
+from __future__ import print_function
+import os
+import yaml
+import time
+import string
+import random
+import logging
+import paramiko
+import argparse
+from openstack_connection import create_openstack_vms, handle_failure, delete_vm
+
+CERT_PATH = '~/.cloudify-test-ca'
+LOCAL_INSTALL_CLUSTER = '/tmp/install_cluster'
+LOCAL_INSTALL_CLUSTER_CERTS = '{}/certs'.format(LOCAL_INSTALL_CLUSTER)
+REMOTE_PARENT_DIRECTORY = '/tmp'
+REMOTE_INSTALL_CLUSTER = '{}/install_cluster'.format(REMOTE_PARENT_DIRECTORY)
+
+
+def retry_with_sleep(func, *func_args, **kwargs):
+    retry_count = kwargs.get('retry_count', 10)
+    delay = kwargs.get('delay', 2)
+    for i in range(retry_count):
+        try:
+            return func(*func_args)
+        except Exception as e:
+            if i < retry_count - 1:
+                time.sleep(delay)
+                continue
+            else:
+                raise e
+
+
+def _create_ssh_client_func(hostname, username, key_file):
+    client = paramiko.SSHClient()
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    client.connect(hostname=hostname, username=username, key_filename=key_file)
+    return client
+
+
+def _create_ssh_client(hostname, username, key_file):
+    return retry_with_sleep(_create_ssh_client_func, hostname, username,
+                            key_file)  # waiting for the VM to run
+
+
+def _blocking_exec_command(client, command):
+    stdin, stdout, stderr = client.exec_command(command)
+    exit_status = stdout.channel.recv_exit_status()
+    if exit_status != 0:
+        raise Exception(stdout.read())
+    return stdin, stdout, stderr
+
+
+class VM:
+    def __init__(self, private_ip, public_ip, name, key_path, vm_username):
+        self.private_ip = private_ip
+        self.public_ip = public_ip
+        self.name = name
+        self.key_path = key_path
+        self.username = vm_username
+        self.client = _create_ssh_client(self.public_ip, self.username,
+                                         self.key_path)
+
+    def exec_command(self, command):
+        return _blocking_exec_command(self.client, command)
+
+
+def scp_local_to_remote(key_path, instance, source_path, destination_path):
+    os.system('scp -i {key_path} -o "StrictHostKeyChecking no" -r '
```

Comment:
ideally replace `os.system` with the subprocess module (no need for shell either)

Reply:
(several instances in the PR)

Reply:
Thanks for the comment, what is the subprocess module?